lib/rust/transformer/src/main.rs

```diff
@@ -402,12 +402,39 @@ async fn read_events_s3<'a>(
                 }
             })
         }
-        None => FramedRead::new(reader, LinesCodec::new())
-            .map(|v| match v {
-                Ok(v) => Ok(Value::from(v)),
-                Err(e) => Err(anyhow!(e)),
-            })
-            .boxed(),
+        None => {
+            if r.key.clone().ends_with("csv") {
+                let mut csv_reader = AsyncReader::from_reader(reader);
+                let mut headers: StringRecord = StringRecord::new();
+                { headers = csv_reader.headers().await?.to_owned(); } // do better here
+                let records = csv_reader.into_records();
+
+                records
+                    .map(move |r| match r {  // move is used for headers scope.. do better here too
+                        Ok(r) => {
+                            let mut record = Map::new();
+                            let _ = r.iter().enumerate().map(|(i, field)| {
+                                record.insert(headers.get(i).unwrap().parse().unwrap(), field.parse().unwrap());
+                            });
+                            let json_string = serde_json::Value::from(record).to_string();
```

Comment:
See above, you can simplify this to stream over deserialized values. Then you can just convert e.g. the deserialized `r: HashMap<String, serde_json::Value>>` into a `serde_json::Value` using `serde_json::Map::from_iter` and then the serde Value into a `vrl::Value`, since iirc `vrl::Value` implements `From` for `serde_json::Object` and you can wrap the `Map` in `Object`.

Reply:
value::Value::Object wants a BTreeMap<String, value::Value> as its type, currently working on a way to iterate over the BTreeMap to change the serde_json::Value into a value::Value

Reply:
I think you can convert to json val first right?:

```rust
let json_val: serde_json::Value = serde_json::Value::Object(serde_json::Map::from_iter(r.into_iter()));
Ok(Value::from(json_val))
```

Reply:
ohh interesting, i'll give that a shot in the am